framework/src/modules/chain/synchronizer/fast_chain_switching_mechanism.js

```diff
@@ -14,21 +14,97 @@
 
 'use strict';
 
-class FastChainSwitchingMechanism {
-	constructor({ storage, logger, blocks, slots, dpos, activeDelegates }) {
-		this.storage = storage;
-		this.logger = logger;
+const { BaseSynchronizer } = require('./base_synchronizer');
+const { addBlockProperties } = require('../blocks');
+const { restoreBlocks, deleteBlocksAfterHeight } = require('./utils');
+const {
+	ApplyPenaltyAndAbortError,
+	ApplyPenaltyAndRestartError,
+	AbortError,
+	RestartError,
+} = require('./errors');
+
+class FastChainSwitchingMechanism extends BaseSynchronizer {
+	constructor({
+		storage,
+		logger,
+		channel,
+		slots,
+		blocks,
+		bft,
+		processor,
+		dpos,
+		activeDelegates,
+	}) {
+		super(storage, logger, channel);
 		this.slots = slots;
 		this.dpos = dpos;
 		this.blocks = blocks;
+		this.bft = bft;
+		this.processor = processor;
 		this.constants = {
 			activeDelegates,
 		};
 		this.active = false;
 	}
 
 	// eslint-disable-next-line class-methods-use-this,no-empty-function
-	async run() {}
+	async run(receivedBlock, peerId) {
+		this.active = true;
+
+		try {
+			const highestCommonBlock = await this._requestLastCommonBlock(peerId);
```

Comment:
this returns `undefined` in some case, so probably we should check that

Reply:
@shuse2 this is checked within the next function